slog.go

```diff
@@ -0,0 +1,81 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2023 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logr
+
+import (
+	"log/slog"
+)
+
+// SlogImplementor is an interface that a logr.LogSink can implement
+// to support efficient logging through the slog.Logger API.
+type SlogImplementor interface {
+	// GetSlogHandler returns a handler which uses the same settings as the logr.LogSink.
+	GetSlogHandler() slog.Handler
+}
+
+// LogrImplementor is an interface that a slog.Handler can implement
+// to support efficient logging through the logr.Logger API.
+type LogrImplementor interface {
+	// GetLogrLogSink returns a sink which uses the same settings as the slog.Handler.
+	GetLogrLogSink() LogSink
+}
+
+// ToSlog returns a slog.Logger which writes to the same backend as the logr.Logger.
+func ToSlog(logger Logger) *slog.Logger {
+	return slog.New(ToSlogHandler(logger))
+}
+
+// ToSlog returns a slog.Handler which writes to the same backend as the logr.Logger.
+func ToSlogHandler(logger Logger) slog.Handler {
+	if slogImplementor, ok := logger.GetSink().(SlogImplementor); ok {
+		handler := slogImplementor.GetSlogHandler()
+		return handler
+	}
+
+	return &slogHandler{sink: logger.GetSink()}
+}
+
+// FromSlog returns a logr.Logger which writes to the same backend as the slog.Logger.
+func FromSlog(logger *slog.Logger) Logger {
+	return FromSlogHandler(logger.Handler())
+}
+
+// FromSlog returns a logr.Logger which writes to the same backend as the slog.Handler.
+func FromSlogHandler(handler slog.Handler) Logger {
+	if logrImplementor, ok := handler.(LogrImplementor); ok {
+		logSink := logrImplementor.GetLogrLogSink()
+		return New(logSink)
+	}
+
+	return New(&slogSink{handler: handler})
+}
+
+func levelFromSlog(level slog.Level) int {
+	if level >= 0 {
+		// logr has no level lower than 0, so we have to truncate.
```

Comment:
logr docs say:

"Negative V-levels have the same meaning as V(0)."

so you don't need to worry about this (minor point)

Reply:
True.

Reply:
At some point I want to discuss with you whether `V(-1)` should be supported, though.

It think it is a valid use case to say "I want *more* output when calling this code here", which is `someCode(logger.V(-1))`. Right now we only support "I want *less* output": `someCode(logger.V(1))`. Ironically, the latter use case is the one that Jordan was concerned about because when seeing `logger.V(5).Info` in `someCode` it's not obvious that running the binary with `-v=5` won't produce that output. That's not a problem with `someCode(logger.V(-1))` and it was a use case that scheduler folks wanted ("produce more output when scheduling a specific pod").

That slog supports "more important than INFO" (= `slog.LevelWarn` = 4) and "less important than INFO" (= `slog.LevelDebug` = -4) while logr only supports "less important than INFO (= `V(4)`) is another aspect to consider. I understand that this was intentional, but it keeps coming up.

Reply:
Agreed. I got carried away when you mentioned negative levels :sweat_smile: